chalice/package.py

```diff
@@ -530,15 +559,272 @@ def _register_cfn_resource_name(self, name):
+
+        template['resource'].setdefault(
+            'aws_cloudwatch_event_rule', {})[
+                resource.resource_name] = {
+                    'name': resource.resource_name,
+                    'schedule_expression': resource.schedule_expression
+        }
+        template['resource'].setdefault(
+            'aws_cloudwatch_event_target', {})[
+                resource.resource_name] = {
+                    'rule': '${aws_cloudwatch_event_rule.%s.name}' % (
+                        resource.resource_name),
+                    'target_id': resource.resource_name,
+                    'arn': self._fref(resource.lambda_function)
+        }
+        template['resource'].setdefault(
+            'aws_lambda_permission', {})[
+                resource.resource_name] = {
+                    'function_name': resource.lambda_function.function_name,
+                    'action': 'lambda:InvokeFunction',
+                    'principal': 'events.amazonaws.com',
+                    'source_arn': "${aws_cloudwatch_event_rule.%s.arn}" % (
+                        resource.resource_name)
+        }
+
+    def _generate_lambdafunction(self, resource, template):
+        # type: (models.LambdaFunction, Dict[str, Any]) -> None
+
+        func_definition = {
+            'function_name': resource.function_name,
+            'runtime': resource.runtime,
+            'handler': resource.handler,
+            'memory_size': resource.memory_size,
+            'tags': resource.tags,
+            'timeout': resource.timeout,
+            'source_code_hash': '${filebase64sha256("%s")}' % (
+                resource.deployment_package.filename),
+            'filename': resource.deployment_package.filename}
+
+        if resource.security_group_ids and resource.subnet_ids:
+            func_definition['vpc_config'] = {
+                'subnet_ids': resource.subnet_ids,
+                'security_group_ids': resource.security_group_ids
+            }
+        if resource.reserved_concurrency is not None:
+            func_definition['reserved_concurrent_executions'] = (
+                resource.reserved_concurrency
+            )
+        if resource.environment_variables:
+            func_definition['environment'] = {
+                'variables': resource.environment_variables
+            }
+        if resource.layers:
+            func_definition['layers'] = list(resource.layers)
+
+        if isinstance(resource.role, models.ManagedIAMRole):
+            func_definition['role'] = '${aws_iam_role.%s.arn}' % (
+                resource.role.resource_name)
+        else:
+            # resource is a PreCreatedIAMRole.
+            role = cast(models.PreCreatedIAMRole, resource.role)
+            func_definition['role'] = role.role_arn
+
+        template['resource'].setdefault('aws_lambda_function', {})[
+            resource.resource_name] = func_definition
+
+    def _generate_restapi(self, resource, template):
```

Comment:
Looking over this again. I just remembered... Did terraform add support for websockets? Right now if you try this on a websocket app you get this traceback:
```
$ chalice package --pkg-format=terraform terraform
Creating deployment package.
Traceback (most recent call last):
  File "/Users/kyleknap/GitHub/chalice/chalice/cli/__init__.py", line 468, in main
    return cli(obj={})
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/kyleknap/.pyenv/versions/3.6.1/envs/oscon/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/kyleknap/GitHub/chalice/chalice/cli/__init__.py", line 397, in package
    packager.package_app(config, out, stage)
  File "/Users/kyleknap/GitHub/chalice/chalice/package.py", line 842, in package_app
    template = self._templater.generate(resources)
  File "/Users/kyleknap/GitHub/chalice/chalice/package.py", line 582, in generate
    self.dispatch(resource, template)
  File "/Users/kyleknap/GitHub/chalice/chalice/package.py", line 85, in dispatch
    handler(resource, template)
  File "/Users/kyleknap/GitHub/chalice/chalice/package.py", line 109, in _default
    raise UnsupportedFeatureError(resource)
chalice.package.UnsupportedFeatureError: WebsocketAPI(resource_name='websocket_api', name='websockets-dev-websocket-api', api_gateway_stage='api', routes=['$default'], connect_function=None, message_function=LambdaFunction(resource_name='websocket_message', function_name='websockets-dev-websocket_message', deployment_package=DeploymentPackage(filename='/Users/kyleknap/GitHub/chalice/sampleapps/websockets/.chalice/deployments/c28cddae9295a6d0d8b267751df45cd9-python3.6.zip'), environment_variables={}, runtime='python3.6', handler='app.message', tags={'aws-chalice': 'version=1.9.1:stage=dev:app=websockets'}, timeout=60, memory_size=128, role=ManagedIAMRole(resource_name='default-role', role_name='websockets-dev', trust_policy={'Version': '2012-10-17', 'Statement': [{'Sid': '', 'Effect': 'Allow', 'Principal': {'Service': 'lambda.amazonaws.com'}, 'Action': 'sts:AssumeRole'}]}, policy=AutoGenIAMPolicy(document={'Version': '2012-10-17', 'Statement': [{'Effect': 'Allow', 'Action': ['logs:CreateLogGroup', 'logs:CreateLogStream', 'logs:PutLogEvents'], 'Resource': 'arn:aws:logs:*:*:*'}, {'Effect': 'Allow', 'Action': ['execute-api:ManageConnections'], 'Resource': 'arn:aws:execute-api:*:*:*/@connections/*'}]}, traits=set())), security_group_ids=[], subnet_ids=[], reserved_concurrency=None, layers=[]), disconnect_function=None)
```
It might be work explicitly throwing an error for this in order to give a friendlier error like we do for cloudformation generation of lambda functions that have s3 events.

Reply:
web sockets support is still pending for terraform split across a half-dozen+ pull requests 
https://github.com/terraform-providers/terraform-provider-aws/issues/7004

noted re specific error message here.